src/Data/URI.php

```diff
@@ -0,0 +1,283 @@
+		$this->fragment = $this->digestFragment(parse_url($uri_string,PHP_URL_FRAGMENT));
+	}
+
+	/**
+	 * Check schema formating. Return it in case of success.
+	 *
+	 * @param	string	$schema
+	 * @throws	\InvalidArgumentException
+	 * @return	string
+	 */
+	protected function digestSchema($schema)
+	{
+		return $this->checkCorrectFormatOrThrow(self::SCHEMA, (string)$schema);
+	}
+
+	/**
+	 * Check host formating. Return it in case of success.
+	 *
+	 * @param	string	$host
+	 * @throws	\InvalidArgumentException
+	 * @return	string
+	 */
+	protected function digestHost($host)
+	{
+		return $this->checkCorrectFormatOrThrow(self::HOST, (string)$host);
+	}
+
+	/**
+	 * Check port formating. Return it in case of success.
+	 *
+ 	 * @param	int	$port
+	 * @throws	\InvalidArgumentException
+	 * @return	int|null
+	 */
+	protected function digestPort($port)
+	{
+		if($port === null)
+		{
+			return null;
+		}
+		assert('is_int($port)');
+		return (int)$this->checkCorrectFormatOrThrow(self::PORT, (string)$port);
+	}
+
+	/**
+	 * Check path formating. Return it in case of success.
+	 *
+	 * @param	string	$path
+	 * @throws	\InvalidArgumentException
+	 * @return	string|null
+	 */
+	protected function digestPath($path)
+	{
+		if($path === null)
+		{
+			return null;
+		}
+		assert('is_string($path)');
+		$path = trim($this->checkCorrectFormatOrThrow(self::PATH, $path),self::PATH_DELIM);
+		if($path !== '')
+		{
+			return $path;
+		}
+	}
+
+	/**
+	 * Check query formating. Return it in case of success.
+	 *
+	 * @param	string	$query
+	 * @throws	\InvalidArgumentException
+	 * @return	string|null
+	 */
+	protected function digestQuery($query)
+	{
+		if($query === null)
+		{
+			return null;
+		}
+		assert('is_string($query)');
+		return $this->checkCorrectFormatOrThrow(self::QUERY, $query);
+	}
+
+	/**
+	 * Check fragment formating. Return it in case of success.
+	 *
+	 * @param	string	$fragment
+	 * @throws	\InvalidArgumentException
+	 * @return	string|null
+	 */
+	protected function digestFragment($fragment)
+	{
+		if($fragment === null)
+		{
+			return null;
+		}
+		assert('is_string($fragment)');
+		return $this->checkCorrectFormatOrThrow(self::FRAGMENT, $fragment);
+	}
+
+
+	/**
+	 * Check wether a string fits a regexp. Return it, if so,
+	 * throw otherwise.
+	 *
+	 * @param	string	$regexp
+	 * @param	string	$string
+	 * @throws	\InvalidArgumentException
+	 * @return	string|null
+	 */
+	protected function checkCorrectFormatOrThrow($regexp,$string)
+	{
+		if(preg_match($regexp, (string)$string) === 1) {
+			return $string;
+		} else {
+			throw new \InvalidArgumentException('ill-formated component '.$string);
+		}
+	}
+
+	/**
+	 * @return	string
+	 */
+	public function schema()
+	{
+		return $this->schema;
+	}
+
+	/**
+	 * @return	string
+	 */
+	public function authority()
+	{
+		$port = $this->port();
+		if($port === null) {
+			return $this->host();
+		}
+		return $this->host().':'.$port;
+
+	}
+
+	/**
+	 * @return	string|null
+	 */
+	public function port()
```

Comment:
I am missing a verb here `get`, `fetch` or a comparable word . Some of these methods can be miss interpreted because some of the nouns(like `port` and `host`) are actually also verbs(Especially when written lower case).

Reply:
I see your issue. I (personally) don't agree with the conclusion that adding `get` is really necessary.

To me the "get-less" properties make it clear, that these really are properties of some data-object and not some operations to perform. Reading something like `$http->redirect($url->host())` to me makes it clear, that I'm only accessing some property of `$url`, while `$http->redirect($url->getHost())` could also indicate some hidden operation, especially compared to `$http->redirect($setting->getHost())`. So I would use verbs only where some substantial operations are performed, while a call to `port` could basically be optimized away to an according memory access.

I also tend to find code without `get` for Getters to be more readable in general, since `get` often only adds noise without adding additional information. Compare `sprintf($lng->getTxt("error"), $error->getPage(), $error->getMessage(), $error->getCode())` to `sprintf($lng->txt("error"), $error->page(), $error->message(), $error->code())`.

I do realize, though, that this is not the current practice, neither in ILIAS nor in PHP in general. This seems to be some legacy from Java and habit. C# and Python (e.g.?) tend to handle that issue differently. Since other objects in data also do not use `get` (`Result`, `Color`, while `DataSize` does) I will rest that issue here. If someone cheers for my arguments against `get` (or at least signals that she wants to consider it) I will raise that issue with some larger audience. If not I will just adjust `URL`, `Result` and `Color`.

Reply:
I know that [getter and setter are the roots of all evil in OOP](https://www.javaworld.com/article/2073723/core-java/why-getter-and-setter-methods-are-evil.html) and in perspective of [Object Calethenics](https://www.cs.helsinki.fi/u/luontola/tdd-2009/ext/ObjectCalisthenics.pdf) this access on member variables should not even permitted, but some [people](https://williamdurand.fr/2013/06/03/object-calisthenics/#9-no-getters/setters/properties) bend this rule to at least read the member variable.

In Python and similar languages member variables can be immutable which leads to a totally different behavior. You can access the data without the need of completely encapsulate the member variables through methods. PHP on the other hand currently does not support immutable variables, which leads to the fact that we have to use methods to access the variables and encapsulate them from suspicious behavior.

IMHO a method should always says what it **does**. `host`, `port`, `message` is still ambiguous even if its just an data container(btw, which the `URL` class is not, because it parses the given string into separated parts).

The [PHP PSRs](https://www.php-fig.org/psr/) also have a lot of `get`-, `create`-methods in their proposed standards and still have them in their newest drafts.

Code semantics always leads to discussion. Everybody has an opinion on this field. We may need a clear code convention for where and how we should name these methods in the future and may add such a conclusion to https://github.com/ILIAS-eLearning/ILIAS/tree/release_5-3/docs/documentation.

Reply:
My point about having `port` (etc.) is, that they do not exactly do anything, as the object represents some data (and not some service or object in the OOP sense).

But I'm absolutely fine with skipping the discussion (and change other occurences of similar issue to `getXYZ`).

Thx!

@dkloepfer Please create a new PR that performs the `getXYZ`-change in all the `Data`-classes. Thx!